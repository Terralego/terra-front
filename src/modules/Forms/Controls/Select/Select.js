import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { MenuItem, Button, Spinner } from '@blueprintjs/core';
import { Select as BPSelect } from '@blueprintjs/select';
import uuid from 'uuid/v4';

import { preventEnterKeyPress } from '../../../../utils/event';
import translateMock from '../../../../utils/translate';
import formatValues from '../formatValues';
import NoValues from '../NoValues';

const MAX_NON_FILTERABLE_COUNT = 9;

export class Select extends React.Component {
  static propTypes = {
    label: PropTypes.string,
    onChange: PropTypes.func,
    values: PropTypes.arrayOf(
      PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.shape({
          label: PropTypes.string.isRequired,
          value: PropTypes.any.isRequired,
        }),
      ]),
    ).isRequired,
    isSubmissionPrevented: PropTypes.bool,
    translate: PropTypes.func,
    loading: PropTypes.bool,
  };

  static defaultProps = {
    label: '',
    isSubmissionPrevented: true,
    onChange () {},
    translate: translateMock({
      'terralego.forms.controls.select.empty_item': 'All',
      'terralego.forms.controls.select.no_results': 'No results.',
      'terralego.forms.controls.select.select_placeholder': 'Enter a query first',
      'terralego.forms.controls.select.input_placeholder': 'Filter...',
      'terralego.forms.controls.noValues': 'No choice available',
    }),
    loading: false,
  };

  static getDerivedStateFromProps ({ values = [], value }) {
    if (!values.length) {
      return {
        values,
        value: null,
      };
    }

    const formatedValues = formatValues(values);

    const foundValue = value
      ? formatedValues.find(({ value: availableVal }) => value === availableVal)
      : formatedValues[0];

    return {
      values: formatedValues,
      value: foundValue,
    };
  }

  state = {
    values: [],
    value: null,
  };

  uuid = `select-${uuid()}`;

  handleChange = ({ value }) => {
    const { onChange } = this.props;
    onChange(value);
  };

  render () {
    const {
      label,
      placeholder,
      isSubmissionPrevented,
      value: origValue,
      className,
      inline,
      fullWidth,
      translate,
      loading,
      ...props
    } = this.props;

    const { values = [], value } = this.state;
    const { handleChange } = this;

    const displayedValue = value ? value.label : '';

    return (
      <div
        className={classnames(
          'control-container',
          'control-container--select',
          { 'control-container--full-width': fullWidth },
        )}
        onKeyPress={isSubmissionPrevented ? preventEnterKeyPress : null}
        role="presentation"
      >
        <label
          htmlFor={this.uuid}
          className={classnames(
            'control-label',
            { 'control-label--inline': inline },
          )}
        >
          {label}
        </label>
        {loading && <Spinner size={20} />}
        {!loading && !values.length && <NoValues placeholder={translate('terralego.forms.controls.noValues')} />}
        {!loading && (
          <BPSelect
            className={classnames('tf-select', className)}
            popoverProps={{ usePortal: false, boundary: 'viewport', minimal: true }}
            items={values}
            filterable={values.length > MAX_NON_FILTERABLE_COUNT}
            onQueryChange={this.handleQueryChange}
            initialContent={(values && values.length > 250)
              ? <MenuItem disabled text={translate('terralego.forms.controls.select.select_placeholder')} />
              : undefined}
            inputProps={{
              placeholder: translate('terralego.forms.controls.select.input_placeholder'),
            }}
            itemPredicate={
              (query, { label: itemLabel }) => itemLabel.toLowerCase().includes(query.toLowerCase())
            }
            itemRenderer={({
              label: itemLabel, value: itemValue,
            }, {
              handleClick, modifiers: { matchesPredicate, ...modifiers },
            }) => (
              <div
                key={`${itemValue}${itemLabel}`}
                className={classnames({
                  'control-container__item': true,
                  'control-container__item--empty': itemValue === '',
                })}
              >
                <MenuItem
                  {...modifiers}
                  active={value.value === itemValue}
                  onClick={handleClick}
                  text={itemLabel}
                />
              </div>
            )}
            noResults={(
              <MenuItem
                disabled
                text={translate('terralego.forms.controls.select.no_results')}
              />
            )}
            onItemSelect={handleChange}
            {...props}
          >
            <Button
              id={this.uuid}
              text={displayedValue || translate('terralego.forms.controls.select.empty_item')}
              rightIcon="double-caret-vertical"
            />
          </BPSelect>
        )}
      </div>
    );
  }
}

export default Select;
