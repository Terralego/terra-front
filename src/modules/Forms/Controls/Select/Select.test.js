import React from 'react';
import renderer from 'react-test-renderer';
import { mount, shallow } from 'enzyme';

import Select from './Select';

jest.mock('@blueprintjs/select', () => ({
  Select: function BPSelect ({ items }) {
    return (
      <select>
        {items.map(({ label, value }) => (
          <option key={`${value}${label}`} value={value}>{label}</option>
        ))}
      </select>
    );
  },
}));
jest.mock('@blueprintjs/core', () => ({
  Button () { return null; },
  Spinner: () => <p>Spinner</p>,
  MenuItem: function BPMenuItem () { return null; },
  Position: {
    BOTTOM_LEFT: 'foo',
  },
}));

jest.mock('uuid/v4', () => () => 'uuid');

it('should render correctly', () => {
  const tree = renderer.create((
    <Select
      label="Pwout"
      onChange={() => null}
      values={['foo']}
    />
  ));
  expect(tree.toJSON()).toMatchSnapshot();
});

it('should render loading correctly', () => {
  const tree = renderer.create((
    <Select label="Pwout" loading values={[]} />
  ));
  expect(tree.toJSON()).toMatchSnapshot();
});

it('should render correctly with query', () => {
  const tree = renderer.create((
    <Select
      label="Pwout"
      onChange={() => null}
      values={['foo', 'bar', 'foofoo', 'fou', 'fooooo']}
    />
  ));
  tree.getInstance().handleQueryChange('fo');
  expect(tree.toJSON()).toMatchSnapshot();
});

it('should handle change', () => {
  const onChange = jest.fn();
  const instance = new Select({ onChange });
  instance.handleChange({ value: 'foo' });
  expect(onChange).toHaveBeenCalledWith('foo');
});

it('should render Item', () => {
  const wrapper = mount(
    <Select
      values={['foo']}
    />,
  );
  const ItemRenderer = ({ item, ...props }) =>
    wrapper.find('BPSelect').props().itemRenderer(
      item,
      {
        ...props,
        modifiers: { ...props.modifiers || {} },
      },
    );

  expect(shallow(
    <ItemRenderer
      item={{
        label: 'Foo',
        value: 'foo',
      }}
    />,
  ).props().className).toBe('control-container__item');
  expect(shallow(
    <ItemRenderer
      item={{
        label: 'empty',
        value: '',
      }}
    />,
  ).props().className).toBe('control-container__item control-container__item--empty');
});

it('should handle query change', () => {
  const instance = new Select();
  instance.setState = jest.fn();
  instance.handleQueryChange('foo');
  expect(instance.setState).toHaveBeenCalledWith({ query: 'foo' });
});

it('should prevent submit event in parent', () => {
  const wrapper = shallow(<Select values={['foo', 'bar']} isSubmissionPrevented={false} />);
  const div = wrapper.find('.control-container');
  expect(div.props().onKeyPress).toBe(null);
});

it('should update values', () => {
  expect(Select.getDerivedStateFromProps({ })).toBe(null);
  expect(Select.getDerivedStateFromProps({ values: ['foo', 'bar'] })).toEqual({
    values: [{
      value: 'foo',
      label: 'foo',
    }, {
      value: 'bar',
      label: 'bar',
    }],
  });
});
