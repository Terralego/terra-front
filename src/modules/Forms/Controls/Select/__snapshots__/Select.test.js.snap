// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div
  className="control-container control-container--select"
  onKeyPress={[Function]}
  role="presentation"
>
  <label
    className="control-label"
    htmlFor="select-uuid"
  >
    Pwout
  </label>
  <select>
    <option
      value="foo"
    >
      foo
    </option>
  </select>
</div>
`;

exports[`should render correctly with a lot of values 1`] = `
<div
  className="control-container control-container--select"
  onKeyPress={[Function]}
  role="presentation"
>
  <label
    className="control-label"
    htmlFor="select-uuid"
  >
    It's over the max
  </label>
  <span>
    InitialContent
  </span>
</div>
`;

exports[`should render correctly with values 1`] = `
<div
  className="control-container control-container--select"
  onKeyPress={[Function]}
  role="presentation"
>
  <label
    className="control-label"
    htmlFor="select-uuid"
  >
    Pwout
  </label>
  <select>
    <option
      value="foo"
    >
      foo
    </option>
    <option
      value="bar"
    >
      bar
    </option>
    <option
      value="foofoo"
    >
      foofoo
    </option>
    <option
      value="fou"
    >
      fou
    </option>
    <option
      value="fooooo"
    >
      fooooo
    </option>
  </select>
</div>
`;

exports[`should render loading correctly 1`] = `
<div
  className="control-container control-container--select"
  onKeyPress={[Function]}
  role="presentation"
>
  <label
    className="control-label"
    htmlFor="select-uuid"
  >
    Pwout
  </label>
  <p>
    Spinner
  </p>
</div>
`;
